### PR TITLE
Support YAML client configurations for cache [HZ-2834]

### PR DESCRIPTION
**Pull request by gbarnett-hz**
Determines which configuration parser to use dependent on the file extension of the specified configuration. For backwards we use XML parser as the default should the configuration not be YAML.

Fixes https://github.com/hazelcast/hazelcast/issues/21741

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [ ] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [ ] Send backports/forwardports if fix needs to be applied to past/future releases
- [ ] New public APIs have `@Nonnull/@Nullable` annotations
- [ ] New public APIs have `@since` tags in Javadoc

### CODE REVIEW

**Comment by vbekiaris** [comment]
run-lab-run

File: hazelcast/src/main/java/com/hazelcast/cache/impl/HazelcastServerCachingProvider.java [hunk 1]
```diff
@@ -133,8 +134,8 @@ private Config getConfigFromLocation(URI location, ClassLoader classLoader, Stri
     }
 
     private Config getConfig(URL configURL, ClassLoader theClassLoader, String instanceName) throws IOException {
-        Config config = new XmlConfigBuilder(configURL).build()
-                .setClassLoader(theClassLoader);
+        Config config = Config.loadFromFile(new File(configURL.getPath()));
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/HazelcastServerCachingProvider.java:137]: this works fine as long as the URL represents a local `file:` or `classpath:` resource. However it breaks the use case for `http:` / `https:` (and probably `jar:` schemes... so after all it looks like we shouldn't use the `Config.loadFromFile` facility and make a guess based on `.xml`/`.yaml`/`.yml` extension of the URL path.

File: hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractHazelcastCachingProvider.java [hunk 2]
```diff
@@ -386,4 +386,25 @@ private boolean isConfigLocation(URI location) {
...
+        String path = configURL.getPath();
+        if (path.endsWith(".xml")) {
+            return true;
+        }
+
+        if (path.endsWith(".yml") || path.endsWith(".yaml")) {
+            return false;
+        }
+
+        throw new IllegalArgumentException(
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractHazelcastCachingProvider.java:407]: Backward compatibility consideration: in previous versions, having a `.xml` extension was not a requirement for configuration URLs. Now a config URL that does not end with `.xml` will fail with an exception.

So there are two options:
- break non-`.xml` parsing, add a note to release notes so people might notice before upgrading.
- assume XML format for any URL path that does not match `.yaml`/`.yml`. This way we don't risk breaking existing working setups. In any case, if not a valid XML file, parsing will fail a bit further down the road.

wdyt?

**Review by vbekiaris — COMMENTED** [review]
I have one comment we need to address, otherwise looks good

File: hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractHazelcastCachingProvider.java (hunk 2 above)
Comment (gbarnett-hz) [reply hazelcast/src/main/java/com/hazelcast/cache/impl/AbstractHazelcastCachingProvider.java:407]: @vbekiaris ah, your second point seems to be best approach IMO: that will satisfy the new YAML piece while being backwards compatible.

If you're OK with that I will go ahead and make the change.

**Review by vbekiaris — APPROVED** [review]

**Review by zpqrtbnk — APPROVED** [review]
ok with me

**Pull request merged**